src/firefly/java/edu/caltech/ipac/firefly/server/query/ObsCorePackager.java

```diff
@@ -0,0 +1,209 @@
+                        ext_file_name = uniquenessCheck(uniqueFileNames,ext_file_name,extension,fileTypeCounter);
+                        FileInfo fileInfo = new FileInfo(access_url, ext_file_name, 0);
+                        fileInfos.add(fileInfo);
+                    }
+                }
+                else { //access_format is null, so try and get it from the url's Content_Type
+                    extension = getExtFromURL(url);
+                    ext_file_name = uniquenessCheck(uniqueFileNames,ext_file_name,extension,fileTypeCounter);
+                    FileInfo fileInfo = new FileInfo(access_url, ext_file_name, 0);
+                    fileInfos.add(fileInfo);
+                }
+
+            }
+            fileInfos.forEach(fileInfo -> fileInfo.setRequestInfo(HttpServiceInput.createWithCredential(fileInfo.getInternalFilename())));
+        }
+        catch(Exception e) {
+            e.printStackTrace();
+        }
+        return Arrays.asList(new FileGroup(fileInfos, null, 0, "ObsCore Download Files"));
+    }
+
+    // uniquenessCheck checks for unique file names (appending (1), (2), etc. for duplicate file names)
+    // It also appends the extension and returns the final file name created
+    public static String uniquenessCheck(Set<String> uniqueFileNames,String extName,String extension,Map<String,Integer>fileTypeCounter) {
+        if (uniqueFileNames.contains(extName + "." + extension)) {
+            Integer counter = fileTypeCounter.get(extension);
+            if (counter == null) {
+                counter = 1;
+            }
+            extName += "("+counter+")." + extension;
+            fileTypeCounter.put(extension, ++counter);
+        }
+        else {
+            extName += "." + extension;
+        }
+        uniqueFileNames.add(extName);
+        return extName;
+    }
+
+    public static List<FileInfo> parseDatalink(URL url, String prepend_file_name) {
+        List<FileInfo> fileInfos = new ArrayList<>();;
+        try {
+            DataGroup[] groups = VoTableReader.voToDataGroups(url.toString(), false);
+            for (DataGroup dg : groups) {
+                Set<String> uniqueFileNames = new HashSet<>();
+                Map<String,Integer> fileTypeCounter = new HashMap<>();
+                for (int i=0; i < dg.size(); i++) {
+                    String accessUrl = String.valueOf(dg.getData("access_url", i));
+                    String sem = String.valueOf(dg.getData("semantics", i));
+                    String content_type = String.valueOf(dg.getData("content_type", i));
+
+                    if (accessUrl == null || sem == null) {
+                        //if only semantic (sem) is null, accessUrl may still be available, but we won't know which accessUrls ones to pick
+                        continue;
+                    }
+
+                    if (sem.equalsIgnoreCase("#this") || sem.equalsIgnoreCase("#thumbnail" ) ||
```

Comment:
because of the standard and the way semantics work it  would be better to
```java
sem.toLowerCase().endsWith("#this") 
```
you might want to make that a function to make the code more readable, such as 
```java
    private static boolean testSem(String s, String val) {
       return (s!=null && s.toLowerCase().endsWith(val));  
    }
```

it is allowable to have a semantic that is something like  `https://my.edu/myCustomSemanitc#this`